Plugin/InitForSave.php

```diff
@@ -56,41 +54,47 @@ public function __construct(
         $this->request = $request;
     }
 
-
     /**
-     * @param \Magento\Catalog\Controller\Adminhtml\Product\Initialization\Helper $subject
-     * @param \Magento\Catalog\Model\Product $product
-     *
-     * @throws \Magento\Framework\Exception\LocalizedException
-     *
-     * @return \Magento\Catalog\Model\Product
+     * @param Helper $subject
+     * @param Product $product
+     * @return Product
+     * @throws LocalizedException
      */
     public function afterInitialize(Helper $subject, Product $product)
     {
-        if ($downloadable = $this->request->getPost('downloadable')) {
-            if (isset($downloadable['attachment']) && is_array($downloadable['attachment'])) {
-                $product->setDownloadableData($downloadable);
-                $extension = $product->getExtensionAttributes();
+        $downloadable = $this->request->getPost('downloadable');
+
+        if (!empty($downloadable) && (isset($downloadable['attachment']) && is_array($downloadable['attachment']))) {
+            $product->setDownloadableData($downloadable);
 
-                $attachments = [];
-                foreach ($downloadable['attachment'] as $attachmentData) {
-                    if (!$attachmentData ||
-                        (isset($attachmentData['is_delete']) && (bool) $attachmentData['is_delete'])) {
-                        continue;
-                    } else {
-                        $attachments[] = $this->attachmentBuilder->setData(
-                            $attachmentData
-                        )->build(
-                            $this->attachmentFactory->create()
-                        );
-                    }
+            $attachments = [];
+            foreach ($downloadable['attachment'] as $attachmentData) {
+                if (!$attachmentData ||
+                    (isset($attachmentData['is_delete']) && (bool) $attachmentData['is_delete'])) {
```

Comment:
tutaj podobnie, dałbym to w osobną metodę

Reply:
All refactor-like notes will be corrected in separate PR, as whole module needs to be refactored (it is quite old). I will skip it in scope of this PR.